src/Database/LSMTree/Internal/Run.hs

```diff
@@ -192,15 +182,17 @@ fromMutable :: HasFS IO h
             -> RunDataCaching
             -> RefCount
             -> RunBuilder RealWorld (FS.Handle h)
-            -> IO (Run RealWorld (FS.Handle h))
+            -> IO (Run IO (FS.Handle h))
 fromMutable fs hbio runRunDataCaching refCount builder = do
     (runRunFsPaths, runFilter, runIndex, runNumEntries) <-
       Builder.unsafeFinalise fs hbio (runRunDataCaching == NoCacheRunData) builder
-    runRefCount <- newMutVar refCount
+    runRefCounter <- RC.unsafeMkRefCounterN refCount Nothing
     runKOpsFile <- FS.hOpen fs (runKOpsPath runRunFsPaths) FS.ReadMode
     runBlobFile <- FS.hOpen fs (runBlobPath runRunFsPaths) FS.ReadMode
     setRunDataCaching hbio runKOpsFile runRunDataCaching
-    return Run {..}
+    let !r = Run{..}
+    RC.replaceFinaliser runRefCounter (Just $ close fs hbio r)
+    pure r
```

Comment:
I see, you do use `replaceFinalier`. I don't know why my search before missed it. Whatever.

Ok, so I think you can tie this knot without needing `replaceFinaliser`. You just need to use recursive do.

```suggestion
    rec runRefCounter <- RC.unsafeMkRefCounterN refCount (Just $ close fs hbio r)
        let r = Run{..}
    pure $! r
```
This of course needs `{-# LANGUAGE RecursiveDo #-}`

And to be clear, I think this would be a cleaner design than requiring the ref count to have a mutable finaliser.

Reply:
Done!